src/context.rs

```diff
@@ -288,6 +291,15 @@ pub struct InnerContext {
 
     /// True if account has subscribed to push notifications via IMAP.
     pub(crate) push_subscribed: AtomicBool,
+
+    /// [MagicEndpoint] needed for iroh peer channels.
+    pub(crate) endpoint: Mutex<Option<MagicEndpoint>>,
```

Comment:
Where do you expect deadlock?

Reply:
I see that `iroh_endpoint.write()` is called under `iroh_gossip.write()`. Also `iroh_channels.write()` is called under `iroh_gossip.read()`. Now there are no ABBA deadlocks, but it should be documented in which order mutexes/rwlocks must be acquired if it's really necessary to acquire multiple locks at the same time (i haven't read all the code yet, so haven't yet understood all this locking magic)

Reply:
Is it mathematically proven acquiring locks in the same order is sound?

Reply:
If you're about Rust, then there's no such restriction and that's one of the reasons why it doesn't protect from deadlocks. Taking locks in the same order everywhere is one of the ways to protect yourself (not the only of course, i just don't remeber them all). But maybe the code just can be refactored so that we never take multiple locks at the same time and don't need to document anything about locking?

Reply:
At the moment the code is not really functional, taking locks at multiple places, so a simple refactoring is not possible. We are currently trying to get this PR into main for the next release, so I would suggest not to do this kind of refactoring now, as there are a lot of other things needed and my capacities are limited at the moment.